Rework of object types

jgebal opened the pull request.

jgebal committed:
Rework of object types

Pazus commented on source/types/ut_assert_result_base.tps:
```diff
@@ -0,0 +1,13 @@
+  expected varchar2(4000 char),
+  actual   varchar2(4000 char),
+  end_time timestamp with time zone,
+  not instantiable member procedure add_element( self in out nocopy ut_assert_result_base, an_assertion ut_assert_result_base ),
```
I see your point of making base class for an assertion result but I don't think "expected" and "actual" properties should be there, I think they should be moved to the subclass like "ut_compart_assert", it assert's may theck error raised, or something else that can't be represented natively by "expected", "actual" values.

jgebal replied:
You've raised an interesting point.
There might a lot of different assertions that give different outcomes.
Most will compare expected vs actual but some will not (probably).
I'll have a look into RSpec code and try to figure out how it's done there - they have lots of different assertions.
The trouble is: how to separate the assertions from reporting of the assert outcomes?
I'd like to give the reporter full control on how to report on asserts.
If we do not provide consistent assertion results to be interpreted by reporter then the reporting will sneak int the assertions - which is the case in the code prior to my PR.
Something to think about.
Maybe we can stick for now to expected+actual as i have no better idea on how to give full control on formatting to the reporter otherwise.
The only thing i'll do is that i'll move those to the ut_assert_result.
As you've noticed the base class in the proposed form and usage doesn't make much sense.

Pazus commented on source/types/ut_assert_result_base.tps:
```diff
@@ -0,0 +1,13 @@
+  not instantiable member procedure add_element( self in out nocopy ut_assert_result_base, an_assertion ut_assert_result_base ),
+  not instantiable member function get_element( a_position integer ) return ut_assert_result_base,
+  not instantiable member function get_elements_count return integer
+)
```
I can't agree with the methods for the base abstraction. ut_assert_result_base is an element of assertions list for the test, it can't have methods oriented to the LIST, it can only have methods for itself. You can't ask an object for i-th element of the list if it dosn't content list at all and shouldn't contain it by the propose of it's class.

I think this methods shouldn't be in the assertion element base class.

jgebal replied:
Agreed. The methods don't belong to the base class.

Pazus commented:
I have noticed the you inherit ut_assert_result_base by ut_execution_results, I don't think that is the right way to go. For the asserts we are not going to use Composite pattern so there is no need for the to have common parent class. Moreover ut_execution_result class has different purpose, it manages result items, adds them to the list, and calls comething like 'print' for the reporters. The only field the intersect is "result" because is can be set as resulting value for the whole list. And here we have to consider, that such "resulting value" can be set not only for the asserts array for the test, but also as as resulting value between test in a suite, or suites in a grouping suite. I think here we need more general abstraction.

What do you think about having base class ut_execution_result as a class for general "resulting value" used for suite basically and a subclass that contains assert list property to be used in ut_test?

Pazus commented:
Don't you think we should separate changes in to two batches: minor fixes of the current state, and object model rework to clearly see such important changes?

jgebal commented:
Hi @pazus
I agree with most of your comments.
Let me rework my commit into smaller pieces.

Pazus commented:
So we plan to stick with design like this:
assert_result_base_abstract (result, message fields, get_result, is_sucess, get_short_message, get_details methods) -> concret_assert_result(comparison check, not null check, some sql check, anything else that will be implemented further)

Then we have list of assert_result.

execution_result - basic structure for ut_test_object (start_time, end_time, result) (as current ut_execution_result)

Hard question. Do we have to subclass execution_result adding list of assert_result as a property.
1) if we do so, then wу would like that class to have method to define it's result value based on the assert_result list items. But we couldn't do the same for the suite's execution result property as the it doesn't have info about tests of the suite. Then the suite class has to have method to define result value what is opposite to the initial goal. An because of difficulties with circular dependencies in PL/SQL we cant pass suite to the execution_result class as a parameter.

I think we have to leave the design the way it is: ut_test_object should have method to define it's execution_result propery. ut_test_object set's the start_time and end_time and should have method to define the "result" property value. This way we don't subclass execution_result for test and don't put assert_result list as a property of the subclass but rather leave it in the ut_test.

Pazus commented on source/types/ut_test.tpb:
```diff
@@ -70,12 +65,13 @@ create or replace type body ut_test is
       end if;
     
+
+      self.execution_result := ut_assert.get_test_results;
```
Here you overwrite previous value with start_time set.

jgebal replied:
nope - the get_tests_results does not impact the start time

Pazus commented on source/types/ut_test.tpb:
```diff
@@ -87,16 +83,17 @@ create or replace type body ut_test is
         ut_assert.report_error(sqlerrm(sqlcode) || ' ' || dbms_utility.format_error_stack);
         ut_assert.report_error(sqlerrm(sqlcode) || ' ' || dbms_utility.format_error_backtrace);
+
+        self.execution_result := ut_assert.get_test_results;
```
here you also overwrite previously set value of self.execution_result

jgebal replied:
This overwrite would only happen if flow would get into the exception block due to exception after line 70.
There is nothing that could cause that - no overwrite.

jgebal commented on source/types/ut_test.tpb:
```diff
@@ -48,7 +43,7 @@ create or replace type body ut_test is
       $end
     
-      self.execution_result := ut_execution_result();
+      self.execution_result := ut_assert.start_new_test( reporter );
```
`start_new_test` initializes the execution results (setting the start time).
Now that I look at it, this probably should be a **procedure not a function**.
Also the statefulness should probably be ouside of ut_assert to avoid confusion.
So maybe a separate package `ut_results_buffer`?

jgebal committed:
Removed composite from ut_assert_result and ut_assert_result_base.
Renamed ut_assert_result_base to ut_execution_result_base - generic type.
Renamed ut_assert_list into ut_execution_list.
Type ut_exectutuion_result is now a composite that can hold a list of any subclass of abstract class ut_execution_result_base, that is, ut_execution_result or ut_assert_result.
Renamed field "message" to field "name".
Added field "error_message".
Changed procedure ut_test_call_param from procedure to function and refactored ut_test.execute to use that feature.
Moved statefulness from ut_assert into ut_assert_buffer.

jgebal commented:
Hi Pavel, 
The circular dependencies ir Oracle Objects seem to be a design mistake, but there is a nice workaround for that.
- Crate a base class
- Create a collection for base class
- Create a subclass that has field - collection of base class

I've used this approach for [any_data project](https://github.com/jgebal/any_data).
The any_data was aimed to allow casting anything into string (reporting/logging) and grew into something more - ability to compare anything to anything with type check that is more restrictive than in Oracle. This could be ideal for restrictive unit testing, where you compare to result-sets and want to make sure that your test will fail when you compare 1 = '1';

jgebal commented:
I've looked ant RSpec reporters briefly.
What I see as a major difference compared to our approach is that the RSpec reporting interface is focused around the outcome:
- did the thing failed (assertion/test/suite)
- did it pass
- did it errored
  It is not focused around what the thing is (if it is an assert/test/context or describe block)

Pazus commented:
How about skype call tonight? Is 10 CET ok?

Pazus commented on source/types/ut_test.tpb:
```diff
@@ -14,29 +14,23 @@ create or replace type body ut_test is
+       not ut_metadata.resolvable(call_params.owner_name, call_params.object_name, call_params.setup_procedure) then false
+      when call_params.teardown_procedure is not null and
+       not ut_metadata.resolvable(call_params.owner_name, call_params.object_name, call_params.teardown_procedure) then false
+      else true
```
Logic regression. To eliminate multiple returns should be rewritten as:
`return call_params.test_procedure is not null 
and ut_metadata.resolvable(call_params.owner_name, call_params.object_name, call_params.test_procedure) 
and (call_params.setup_procedure is null OR ut_metadata.resolvable(call_params.owner_name, call_params.object_name, call_params.setup_procedure))
and (call_params.teardown_procedure is null OR ut_metadata.resolvable(call_params.owner_name, call_params.object_name, call_params.teardown_procedure));`

Pazus commented:
As most of the suggestions were included in to PR to the main repo I close this PR.
